File: loopx/control_plane/work_items/replan_settlement.ts
import type { JsonObject } from "../effect_program.ts";
import {
  optionalNonEmptyString,
  requireJsonObject,
  requireNonEmptyString,
} from "../runtime_decode.ts";
import {
  isLocalTodoCompletionIdentity,
  localTodoCompletionIdentity,
} from "../todos/completion_fence.ts";

export const REPLAN_SETTLEMENT_REQUEST_SCHEMA =
  "loopx_replan_settlement_request_v0";
export const REPLAN_SETTLEMENT_CONTRACT_SCHEMA =
  "replan_settlement_contract_v0";
export const TODO_LIFECYCLE_REENTRY_REQUEST_SCHEMA =
  "loopx_todo_lifecycle_reentry_request_v0";
export const TODO_LIFECYCLE_REENTRY_SCHEMA =
  "todo_lifecycle_settlement_reentry_v0";

function shellArgument(value: string): string {
  if (/^[A-Za-z0-9_./:@%+=,-]+$/.test(value)) return value;
  return `'${value.replaceAll("'", `'"'"'`)}'`;
}

function argumentSuffix(values: readonly string[]): string {
  return values.length > 0
    ? ` ${values.map(shellArgument).join(" ")}`
    : "";
}

function lifecycleTrigger(value: unknown, index: number): JsonObject {
  const trigger = requireJsonObject(value, `triggers[${index}]`);
  if (trigger.kind !== "completed_advancement_without_successor") {
    throw new Error(`triggers[${index}].kind is unsupported`);
  }
  return {
    kind: trigger.kind,
    todo_id: requireNonEmptyString(
      trigger.todo_id,
      `triggers[${index}].todo_id`,
    ),
    completion_turn_key: optionalNonEmptyString(
      trigger.completion_turn_key,
      `triggers[${index}].completion_turn_key`,
    ),
  };
}

export function projectTodoLifecycleSettlementReentry(value: unknown): JsonObject {
  const request = requireJsonObject(
    value,
    "work_item.replan_settlement.reentry params",
  );
  if (request.schema_version !== TODO_LIFECYCLE_REENTRY_REQUEST_SCHEMA) {
    throw new Error("Todo lifecycle reentry request schema mismatch");
  }
  const goalId = requireNonEmptyString(request.goal_id, "goal_id");
  if (!Array.isArray(request.triggers) || request.triggers.length === 0) {
    throw new Error("triggers must be a non-empty array");
  }
  const triggers = request.triggers.slice(0, 3).map(lifecycleTrigger).map(
    (trigger) => {
      const todoId = String(trigger.todo_id);
      const persistedKey = trigger.completion_turn_key;
      const completionIdentityKey = typeof persistedKey === "string"
        ? persistedKey
        : localTodoCompletionIdentity(goalId, todoId);
      return {
        kind: trigger.kind,
        todo_id: todoId,
        completion_turn_key: completionIdentityKey,
        completion_identity_source: isLocalTodoCompletionIdentity(
            completionIdentityKey,
          )
          ? "unscoped_completion"
          : "turn_settlement",
      };
    },
  );
  const lifecycleActorArgs = Array.isArray(request.lifecycle_actor_args)
    ? request.lifecycle_actor_args.map((item, index) =>
      requireNonEmptyString(item, `lifecycle_actor_args[${index}]`)
    )
    : [];
  const quotaScopedArgs = Array.isArray(request.quota_scoped_args)
    ? request.quota_scoped_args.map((item, index) =>
      requireNonEmptyString(item, `quota_scoped_args[${index}]`)
    )
    : [];
  const nextCliActions = triggers.map((trigger) => {
    const todoId = String(trigger.todo_id);
    const completionIdentityKey = String(trigger.completion_turn_key);
    const completionIdentityArgs =
      trigger.completion_identity_source === "unscoped_completion"
        ? ["--completion-identity-key", completionIdentityKey]
        : ["--turn-instance-id", completionIdentityKey];
    return (
      `when no real successor remains for completed Todo ${todoId}: ` +
      `loopx todo complete --goal-id ${shellArgument(goalId)} ` +
      `--todo-id ${shellArgument(todoId)}` +
      argumentSuffix(completionIdentityArgs) +
      argumentSuffix(lifecycleActorArgs) +
      " --no-follow-up --note '<public-safe no-follow-up rationale>'"
    );
  });
  nextCliActions.push(
    "otherwise link or create one real runnable successor for the exact " +
      "completed Todo; do not create lifecycle-only filler",
    `loopx --format json quota should-run --goal-id ${shellArgument(goalId)}` +
      argumentSuffix(quotaScopedArgs),
  );
  return {
    schema_version: TODO_LIFECYCLE_REENTRY_SCHEMA,
    resolution_mode: "todo_lifecycle_settlement",
    triggers,
    next_cli_actions: nextCliActions,
  };
}

export function projectReplanSettlementContract(value: unknown): JsonObject {
  const request = requireJsonObject(value, "work_item.replan_settlement params");
  if (request.schema_version !== REPLAN_SETTLEMENT_REQUEST_SCHEMA) {
    throw new Error("Replan settlement request schema mismatch");
  }
  const selectedTodoId = optionalNonEmptyString(
    request.selected_todo_id,
    "selected_todo_id",
  );
  const semanticObligationId = requireNonEmptyString(
    request.semantic_replan_obligation_id,
    "semantic_replan_obligation_id",
  );
  const semanticObligationSettlementBound = selectedTodoId === null;
  const settlementBindingKind = semanticObligationSettlementBound
    ? "autonomous_replan"
    : "todo";
  const settlementBindingId = selectedTodoId ?? semanticObligationId;

  return {
    schema_version: REPLAN_SETTLEMENT_CONTRACT_SCHEMA,
    single_binding_required: true,
    settlement_binding: {
      kind: settlementBindingKind,
      id: settlementBindingId,
      cli_argument: semanticObligationSettlementBound
        ? "--replan-obligation-id"
        : "--todo-id",
    },
    semantic_obligation: {
      kind: "autonomous_replan",
      id: semanticObligationId,
      settlement_bound: semanticObligationSettlementBound,
      discharge: semanticObligationSettlementBound
        ? "direct_settlement"
        : "todo_bound_writeback",
    },
  };
}

File: loopx/control_plane/work_items/replan_settlement.py
"""Python adapter for the TypeScript-owned replan settlement projection."""

from __future__ import annotations

from collections.abc import Mapping
from typing import Any

from ..effect_runtime import EffectRuntimeRejected, effect_runtime_result

REPLAN_SETTLEMENT_REQUEST_SCHEMA = "loopx_replan_settlement_request_v0"
REPLAN_SETTLEMENT_CONTRACT_SCHEMA = "replan_settlement_contract_v0"
TODO_LIFECYCLE_REENTRY_REQUEST_SCHEMA = "loopx_todo_lifecycle_reentry_request_v0"
TODO_LIFECYCLE_REENTRY_SCHEMA = "todo_lifecycle_settlement_reentry_v0"


def project_todo_lifecycle_settlement_reentry(
    *,
    goal_id: str,
    triggers: list[dict[str, Any]],
    lifecycle_actor_args: list[str],
    quota_scoped_args: list[str],
) -> dict[str, Any]:
    """Return TS-owned exact actions for a terminal Todo succession gap."""

    try:
        result = effect_runtime_result(
            "work_item.replan_settlement.reentry",
            {
                "schema_version": TODO_LIFECYCLE_REENTRY_REQUEST_SCHEMA,
                "goal_id": goal_id,
                "triggers": triggers,
                "lifecycle_actor_args": lifecycle_actor_args,
                "quota_scoped_args": quota_scoped_args,
            },
        )
    except EffectRuntimeRejected as exc:
        raise ValueError(str(exc)) from None
    if not isinstance(result, Mapping):
        raise RuntimeError("TypeScript Todo lifecycle reentry result must be an object")
    projected_triggers = result.get("triggers")
    next_cli_actions = result.get("next_cli_actions")
    expected_triggers = triggers[:3]
    if (
        result.get("schema_version") != TODO_LIFECYCLE_REENTRY_SCHEMA
        or result.get("resolution_mode") != "todo_lifecycle_settlement"
        or not isinstance(projected_triggers, list)
        or not projected_triggers
        or any(
            not isinstance(trigger, Mapping)
            or trigger.get("kind") != "completed_advancement_without_successor"
            or not isinstance(trigger.get("todo_id"), str)
            or not trigger.get("todo_id")
            or not isinstance(trigger.get("completion_turn_key"), str)
            or not trigger.get("completion_turn_key")
            or trigger.get("completion_identity_source")
            not in {"turn_settlement", "unscoped_completion"}
            for trigger in projected_triggers
        )
        or any(
            dict(projected).get("kind") != expected.get("kind")
            or dict(projected).get("todo_id") != expected.get("todo_id")
            or (
                expected.get("completion_turn_key") is not None
                and dict(projected).get("completion_turn_key")
                != expected.get("completion_turn_key")
            )
            or (
                expected.get("completion_turn_key") is None
                and not str(
                    dict(projected).get("completion_turn_key") or ""
                ).startswith("local_completion_")
            )
            or dict(projected).get("completion_identity_source")
            != (
                "unscoped_completion"
                if expected.get("completion_turn_key") is None
                or str(expected.get("completion_turn_key") or "").startswith(
                    "local_completion_"
                )
                else "turn_settlement"
            )
            for projected, expected in zip(
                projected_triggers,
                expected_triggers,
                strict=True,
            )
        )
        or not isinstance(next_cli_actions, list)
        or not next_cli_actions
        or any(not isinstance(action, str) or not action for action in next_cli_actions)
    ):
        raise RuntimeError("TypeScript Todo lifecycle reentry result shape mismatch")
    return dict(result)


def project_replan_settlement_contract(
    *,
    selected_todo_id: str | None,
    semantic_replan_obligation_id: str,
) -> dict[str, Any]:
    """Return the one causal binding for a semantic replan writeback."""

    try:
        result = effect_runtime_result(
            "work_item.replan_settlement.project",
            {
                "schema_version": REPLAN_SETTLEMENT_REQUEST_SCHEMA,
                "selected_todo_id": selected_todo_id,
                "semantic_replan_obligation_id": semantic_replan_obligation_id,
            },
        )
    except EffectRuntimeRejected as exc:
        raise ValueError(str(exc)) from None
    if not isinstance(result, Mapping):
        raise RuntimeError("TypeScript replan settlement result must be an object")
    binding = result.get("settlement_binding")
    obligation = result.get("semantic_obligation")
    if (
        result.get("schema_version") != REPLAN_SETTLEMENT_CONTRACT_SCHEMA
        or result.get("single_binding_required") is not True
        or not isinstance(binding, Mapping)
        or binding.get("kind") not in {"todo", "autonomous_replan"}
        or not isinstance(binding.get("id"), str)
        or binding.get("cli_argument") not in {"--todo-id", "--replan-obligation-id"}
        or not isinstance(obligation, Mapping)
        or obligation.get("kind") != "autonomous_replan"
        or not isinstance(obligation.get("id"), str)
        or not isinstance(obligation.get("settlement_bound"), bool)
        or obligation.get("discharge")
        not in {"direct_settlement", "todo_bound_writeback"}
    ):
        raise RuntimeError("TypeScript replan settlement result shape mismatch")
    return dict(result)
